usr/src/uts/common/qqcache/qqcache.c

```diff
@@ -0,0 +1,471 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright 2020 Joyent, Inc.
+ */
+
+#include <sys/debug.h>
+#include <sys/errno.h>
+#include <sys/null.h>
+#include <sys/types.h>
+#include <sys/qqcache.h>
+#include <sys/qqcache_impl.h>
+#include <sys/stddef.h>
+
+/*
+ * Currently, the non _KERNEL pieces are to support testing in usr/src/test.
+ */
+#ifdef _KERNEL
+#include <sys/kmem.h>
+#define	ZALLOC kmem_zalloc
+#define	FREE kmem_free
+#else
+#include <umem.h>
+#define	ZALLOC umem_zalloc
+#define	FREE umem_free
+#endif
+
+
+/*
+ * The *_overflow functions mimic the gcc/clang intrinsic functions.  Once
+ * we are using a newer compiler version to that includes these as intrisnics,
+ * these can be replaced with those versions.
+ */
+static int
+uadd_overflow(const size_t a, const size_t b, size_t *sump)
+{
+	*sump = a + b;
+	if (*sump < a || *sump < b)
+		return (1);
+	return (0);
+}
+
+#define	MUL_NO_OVERFLOW ((size_t)1 << (sizeof (size_t) * 4))
+
+static int
```

Comment:
It was intended to match the gcc intrinsic signatures. The hope was to eventually replace with the gcc intrinsics (clang also has them). Since this might be a candidate for upstreaming, unfortunately, with sparc still on gcc-4.4.4 we can't use the intrinsics yet (they came in in 4.4.8).

Reply:
Ack.